pkg/agent/controller/egress/ip_assigner_linux.go

```diff
@@ -0,0 +1,185 @@
+	if err := netlink.AddrAdd(a.egressLink, &addr); err != nil {
+		return fmt.Errorf("failed to add ip %v to interface %s: %v", addr, ifaceName, err)
+	}
+	if err := a.saveIPAssignFile(egressIP, egressName); err != nil {
+		return fmt.Errorf("failed to save egress ip assign, egress %s, ip: %s: %v", egressName, egressIP, err)
+	}
+	isIPv4 := egressSpecIP.To4()
+	if isIPv4 != nil {
+		if err := arping.GratuitousARPOverIface(isIPv4, a.egressInterface); err != nil {
+			klog.ErrorS(err, "Failed to send gratuitous ARP")
+			return err
+		}
+		klog.InfoS("Sent gratuitous ARP", "ip", isIPv4)
+	} else if isIPv6 := egressSpecIP.To16(); isIPv6 != nil {
+		klog.ErrorS(ipv6NotSupportErr, "Failed to send Advertisement")
+		return ipv6NotSupportErr
+	}
+	klog.V(2).InfoS("Added ip", "ip", addr, "interface", ifaceName)
+	return nil
+}
+
+// UnassignEgressIP unassign an egressIP and delete persistent files
+func (a *ipAssigner) UnassignEgressIP(egressIP string) error {
+	egressSpecIP := net.ParseIP(egressIP)
+	addr := netlink.Addr{IPNet: &net.IPNet{IP: egressSpecIP, Mask: net.CIDRMask(32, 32)}}
+	ifaceName := a.egressLink.Attrs().Name
+	if err := netlink.AddrDel(a.egressLink, &addr); err != nil {
+		return fmt.Errorf("failed to delete ip %v from interface %s: %v", addr, ifaceName, err)
+	}
+	if err := a.removeIPAssignFile(egressIP); err != nil {
+		return fmt.Errorf("failed to remove egress ip assign, egressIP: %s: %v", egressIP, err)
+	}
+	klog.V(2).InfoS("Deleted ip", "ip", addr, "interface", ifaceName)
+	return nil
+}
+
+// IsAssignedIP check if a IP address has been assigned with specific egressName
+func (a *ipAssigner) IsAssignedIP(egressName string) (egressIP string, has bool) {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+	if ip, ok := a.assignedEgress[egressName]; ok {
+		return ip, true
+	}
+	return
+}
+
+func ipSavedFile(dir, name string) string {
+	return filepath.Join(dir, name)
+}
+
+func (a *ipAssigner) saveIPAssignFile(egressIP, egressName string) error {
+	var buffer bytes.Buffer
+	buffer.WriteString(egressIP)
+	f, err := os.Create(ipSavedFile(a.egressRunDir, egressName))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = io.Copy(f, &buffer)
+	if err == nil {
+		a.mutex.Lock()
+		defer a.mutex.Unlock()
+		a.assignedEgress[egressName] = egressIP
+		return nil
+	}
+	return err
+}
+
+func (a *ipAssigner) removeIPAssignFile(egressIP string) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	for egressName, ip := range a.assignedEgress {
+		if ip == egressIP {
+			fileName := ipSavedFile(a.egressRunDir, egressName)
+			err := os.Remove(fileName)
+			if err != nil {
+				return err
+			}
+			delete(a.assignedEgress, egressName)
+			return nil
+		}
+	}
+	klog.Warningf("Remove egressIP assignment, egressIP %s not found", egressIP)
+	return nil
+}
+
+// AssignedIPs return assigned a map of [egressName]IP
+func (a *ipAssigner) AssignedIPs() map[string]string {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+	return a.assignedEgress
```

Comment:
Caller is still not thread-safe to use the returned map.

Reply:
done